rfcs/rdm-0079-collections.md

```diff
@@ -0,0 +1,337 @@
+### Design
+    
+At a glance, a collection can be thought of as a set of records. To do so, we further expand the definition into:
+    
+- A filter to obtain the said records from search.
+- A dedicated page that displays only the records that belong to the collection
+    
+Collections can be used for all records, or records in a  community, which has opted in, like the EU Open Research Repository.
+    
+See the example below of a collection:
+
+![](0079/2.png)
+    
+In this example, the collection "Horizon 2020" filters the records from the community "EU" by their funding programme being "Horizon 2020", and the collection page clearly displays the results in a dedicated page inside the European Repository community.
+
+Collections can be configured per community. I.e. for the European Repository it is interesting to have collections based on the records funded by the European Comission, while for a Biodiversity Repository it could be interesting to have these collections on different species (e.g. ants, insects, etc).
+
+For instance, see below multiple collections that exist under the community "European Open Repository":
+
+![](0079/1.png)
+    
+### Technical implementation
+
+Naturally, collections are organized in tree-like structures. I.e. Cern Document Server (CDS) defines multiple top-level collections [1] that each have, potentially, other collections attached.
+    
+![1 - Cern Document Server collections example](0079/3.png)
+*[1] - CDS collections page, taken from https://cds.cern.ch/*
+    
+This can be represented in a tree (for simplicity only three levels are shown):
+    
+![](0079/4.png){width=1000}
+
+Another example of collections:
+    
+![](0079/5.png){width=500}
+*Euro Science subjects collections - http://publications.europa.eu/resource/dataset/euroscivoc*
+
+This solution proposes collections to be modelled as generic trees, whose information is persisted in a database. 
+    
+That said, collections data model will be based in the "Materialized Path" pattern. Read this [documentation from MongoDB](https://www.mongodb.com/docs/manual/tutorial/model-tree-structures-with-materialized-paths/) for a good overview of this pattern. The main concept is that each node (row in the DB) stores its `path` from the root to itself, excluding. For our example, it looks as follows:
+
+
+In our proposed model, however, a unique `id` is to be stored in the `path` so the ascendants of a node can be derived from the path and done with primary key lookups:
+    
+```jsonld
+   { _id: 1, title: "Natural Sciences", path: "," },
+   { _id: 2, title: "Physical science", path: ",1," },
+   { _id: 3, title: "Mathematics", path: ",1," },
+   { _id: 4, title: "Particles", path: ",1,2," },
+```
+
+**Database tables**
+
+We identified three new tables for this design:
+                                     
+1) `Collection`
+2) `CollectionTree`
+3) `CommunityCollectionTree`
+                                     
+![](0079/6.png)
```

Comment:
from this diagram I understand we can have
- many collection trees ( I understand that it is for community purposes, right?)
- one collection can belong to only one tree ---> meaning we might have duplicated objects in different trees, are we OK with that?
- one community can have many collection trees -> what use case does it fulfill ? why not one to one?

Reply:
thanks for the answers, I have follow-up questions:

2. by duplicated objects I meant we might have a collection with query "subjects:bla"  under community EU and "subjects:bla" collection under community BLR (each collection with different id, but the query is the same). My question was if we are aware of this and accepting it.
3. have you considered the community query to be the root collection and the grouping is simply lvl 2 of the tree?

Reply:
@alejandromumo we discussed point 2, we can close it.
However, n. 3 is a valid point, which was also already discussed in the past. What if a community does not need a `tree`, but would simply like to create a set of collections?
We could have a "default" tree, and maybe hide from the UI, however, in the URL, the tree slug is required. I think that this deserves a quick chat.

Reply:
Currently, a `tree` is mandatory to group the collections. But indeed we could have a `default` tree per community, so users don't have to do it every time they need a collection.

It could be e.g. `/communities/<slug>/collections/default/<collection>` where `default` is the `tree.slug` 

WDYT? ping @lnielsen

Reply:
Let's discuss IRL tomorrow. If I understand correctly, the main issues is the URL pattern if you only have one collection tree in a community? I.e. currently you need ``/communities/<id>/collections/<tree-id>/<collection-slug>`` vs you would like to remove ``tree-id`` so it becomes ``/communities/<id>/collections/<collection-slug>``

My five cents is I agree it would be nice to avoid the tree id when only one tree exists, and possibly we could simply allow the tree slug to be empty (or mark the tree as the default one). I makes the resolution of a URL a little bit more complex and there's some consistency checks we would need to take care of (e.g. ensuring that a tree id and community slug is not overlapping)